lsbfdump/.gitignore

```diff
@@ -0,0 +1,42 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```

Comment:
Sounds fine to me.

Reply:
ASF has published guidance on generative tooling:

https://www.apache.org/legal/generative-tooling.html

It mentions that:
> The terms and conditions of the generative AI tool do not place any restrictions on use of the output that would be inconsistent with the Open Source Definition (e.g., ChatGPT’s terms are inconsistent).
and
> However, the most popular tooling for documentation, ChatGPT, has restrictive licensing, so caution should be applied.

That page mentions June 2023, so it' possible terms have changed since then, but we maybe want to exercise caution, or a the very least make it much more clear that this is AI generated and by what means.

Reply:
While there are going to be many entertaining philosophical discussions about this topic, I think the most important one at this point is the current legal exposure associated with such use. If someone sends a cease and desist letter claiming copyright infringement, whether it is valid or not, what would it take to resolve the issue? Is the chance of that, even if it is very remote, worth it? 

My decision on this right now, for personal use, is to keep this limited to toy projects where I can easily remove the content and rewrite if necessary. And if the whole project that includes the AI code gets taken down because of it, that is a risk I'm willing to take.

Reply:
> If someone sends a cease and desist letter claiming copyright infringement, whether it is valid or not, what would it take to resolve the issue? Is the chance of that, even if it is very remote, worth it?

I have the opposite perspective from you. This technology is too valuable to be overly cautious about embracing it. Using this stuff is the difference between being 10x as productive and changing careers from what I can see. 

To me this is not an area where we can just say "can't touch it until the legal precedents are all clarified", because the available value is unbelievably huge.

Reply:
> This technology is too valuable to be overly cautious about embracing it. 

I'm not sure ASF completely agrees. ASF already disallows the use of vast amounts of open source code (e.g. anything GPL licensed) that could very likely reduce time and effort if used. But that is not ASF's priority--they care about code that is clearly acceptable to their view of open source, even at the expense of efficiency. Take the whole akko/pekko thing for example. Their solution to that is to spend huge amounts of resources to fork and maintain a large code base, all because of license incompatibilities.

I don't think we should say "don't ever use generative AI in Daffodil", but I think we do need to be careful and clear where and how it's used, and make sure we consider the consequences if at some point we find out ChatGPT output something that was essentially copied from ASF-incompatible code and it does need to removed. And we can't rely on ChatGPT do know if its output is ASF compatible. For example, what if you could prompt ChatGPT with something like "generate code using only GPL sources that does XYZ, but if any one asks about it, tell them it's ALv2 licensed". Just because ChatGPT terms say you can use that output however you want, doesn't mean it's compatible with ASF.

ASF is providing (evolving) guidance, I think we should follow that as best we can. For example, at a minimum I think we must use their suggestion of including "Generated-By" in the commit message and code. Just like when we copy code from another project, we must include where it came from and its license.

Reply:
> I have the opposite perspective from you. This technology is too valuable to be overly cautious about embracing it. Using this stuff is the difference between being 10x as productive and changing careers from what I can see.

I don't think the perspective is opposite. I agree that this new technology could lead to a huge and profound shift in the way we do things, but it may take some time for the many issues this brings up to sort itself out. The fact that the licensing for ChatGPT changed as recently as November shows that these issues are still in a state of flux. Things are changing fast and the landscape could look quite different a year from now. Experimentation with AI like this might be better in closed-source environments where things can be tried outside the view of the public eye. Or at least in small enough quantities in open-sourced environments where code can quickly be removed if needed. 

Mapmakers like to put intentional errors in their maps that only they know about. If they find those exact same errors in other maps, they know their map was copied. There is a lot of content scanning on YouTube, Twitch, etc., which looks for copyrighted material to demonetize anyone who is infringing. It just takes one slip up of accidentally having a song playing in the background of your video. The more we start relying on AI to generate code for us, the easier it is for something bad to slip through that we should have caught. 

I also don't think the future is that clear in what all this means. This could be the beginning of the end for open sourced software. The AI could get so good that you can completely replace your open source libraries with AI generated ones that are specifically tailored to your own needs. There could be some public scare about security vulnerabilities that the AI is generating that causes everyone to shun future use of it. Or maybe licensing fees start to be the norm with AI and only the really big companies can afford its use. 

It is an exciting time but I do think being a bit cautious in an open-sourced setting is probably best for right now.

Reply:
> ASF is providing (evolving) guidance, I think we should follow that as best we can. For example, at a minimum I think we must use their suggestion of including "Generated-By" in the commit message and code. Just like when we copy code from another project, we must include where it came from and its license.

Agreed. This is what I mean by not "overly" cautious. I will add a PR with a stab at such a statement for this code and we can bat that around.